swim/join_delayer.go

```diff
@@ -0,0 +1,204 @@
+// Copyright (c) 2015 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package swim
+
+import (
+	"errors"
+	"math"
+	"math/rand"
+	"time"
+
+	"github.com/uber-common/bark"
+)
+
+var errAttemptNumberInvalid = errors.New("joinDelayer requires attemptNumber to be greater or equal to 0")
+var errLoggerRequired = errors.New("joinDelayer requires a logger")
+
+const (
+	defaultInitial = float64(100 * time.Millisecond)
+	defaultMax     = float64(60 * time.Second)
+)
+
+var delayerRand = rand.New(rand.NewSource(time.Now().UnixNano()))
+var defaultRandomizer = delayerRand.Intn
+var defaultSleeper = time.Sleep
+var noDelay = time.Duration(0)
+
+// joinDelayer defines the API of a delayer implementation that
+// applies a delay in between repeated join attempts.
+type joinDelayer interface {
+	delay(attemptNumber int) (time.Duration, error)
+}
+
+// delayRandomizer is a function that returns a random number between
+// 0 and an upper-bound, provided in its first argument.
+type delayRandomizer func(int) int
+
+// delaySleeper is a function that pauses execution for time.Duration.
+type delaySleeper func(time.Duration)
+
+// delayOpts is a struct that houses substitutable parameters of a joinDelayer
+// in order to alter its behavior.
+type delayOpts struct {
```

Comment:
We use the full word Options in other places in the code. And would configuration be a more correct name?